### DISCUSSION TITLE
How are folks using this with Java?

### DISCUSSION THREAD

**Discussion by dsabsay** [prabirshrestha/vim-lsp#1348] [Q&A]
I was able to get the Java [jdtls](https://github.com/eclipse/eclipse.jdt.ls) working with the following config:

```
if executable('java') && filereadable(expand('~/lsp/jdt-language-server-1.14.0-202207211651/plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar'))
    au User lsp_setup call lsp#register_server({
                \ 'name': 'eclipse.jdt.ls',
                \ 'cmd': {server_info->[
                \ 'java',
                \ '-Declipse.application=org.eclipse.jdt.ls.core.id1',
                \ '-Dosgi.bundles.defaultStartLevel=4',
                \ '-Declipse.product=org.eclipse.jdt.ls.core.product',
                \ '-Dlog.level=ALL',
                \ '-noverify',
                \ '-Dfile.encoding=UTF-8',
                \ '-Xmx1G',
                \ '--add-modules=ALL-SYSTEM',
                \ '-jar',
                \ expand('~/lsp/jdt-language-server-1.14.0-202207211651/plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar'),
                \ '-configuration',
                \ expand('~/lsp/jdt-language-server-1.14.0-202207211651/config_mac'),
                \ '-data',
                \ getcwd()
                \ ]},
                \ 'whitelist': ['java'],
                \ })
endif
```

While it worked for very simple things like `:LspReferences` or `:LspDefinition` for something in the same file, I found it to be quite limited.

I can't do simple navigation between classes in different packages. For example I have `com.foo.bar.App` that has a method returning a `MyClass` from `com.foo.bar.qux.MyClass`. `:LspDefinition` says it can't find any definitions. Same for `:LspTypeDefinition`. However, if I open `com.foo.bar.qux.MyClass`, those commands start working as expected from the `App` class.

Is there some setting I'm missing? Do I need to somehow tell the language server to load all the files in my project first?

**Comment by prabirshrestha** [comment]
You probably want to enable logs? Are you hitting into this issue https://github.com/prabirshrestha/vim-lsp/pull/1284?

**Reply by dsabsay** [reply]
I'm looking at the logs. It doesn't look like the linked issue is related to this one.
The response coming back from the server doesn't have that `jdt` URI.

Here's a few of the logs (with file paths redacted):

```
Tue Aug  9 17:16:10 2022:["--->",1,"eclipse.jdt.ls",{"method":"textDocument/definition","on_notification":"---funcref---","params":{"textDocument":{"uri":"file:///redacted/long/path/Application.java"},"position":{"character":13,"line":98}}}]
Tue Aug  9 17:16:10 2022:["<---",1,"eclipse.jdt.ls",{"response":{"method":"window/logMessage","jsonrpc":"2.0","params":{"message":"Aug 9, 2022, 5:16:10 PM >> document/definition","type":3}}}]
Tue Aug  9 17:16:10 2022:["<---",1,"eclipse.jdt.ls",{"response":{"id":71,"jsonrpc":"2.0","result":[]},"request":{"id":71,"jsonrpc":"2.0","method":"textDocument/definition","params":{"textDocument":{"uri":"file:///redacted/long/path/Application.java"},"position":{"character":13,"line":98}}}}]
```

Again, after I open the file that contains the definition, the server responds properly with the `file://` URI and Vim loads it.

**Comment by ipkiss42** [comment]
Assuming that `com.foo.qux.MyClass` is not only in a different java package but also a different LSP root folder, then what you are describing is the typical symptom of a multi-root workspace where you haven't told the client about the root folders. Folders get discovered only after you open a class in them, but before that, the language server often has no way to know where else to look for class definitions.

Each LSP client has (or should have) some way to declare the root folders. I'm not sure how it works with vim-lsp, but hopefully others can chime in.

**Reply by dsabsay** [reply]
I wasn't as precise as I should have been. `App` is is in the `com.foo.bar` package and everything I'm trying to resolve is in a separate packages in the same folder as `App`. I've updated my initial post to reflect this.

Your theory seems valid. `getcwd()` gets passed to the server itself and I start vim a few levels higher in the filesystem tree, so every class I mentioned is in a subdirectory of `getcwd()`. But perhaps I need to explicitly give it the folders of all packages? I assumed it would recursively search the cwd, but that doesn't seem to be happening.
